### PR TITLE
ECDb profile version requirements

### PR DESCRIPTION
**Pull request by rschili**
This involved an ECDb profile version upgrade from 4.0.0.3 to 4.0.0.4.

Introducing 2 new custom attributes in ECDbMap which can be used to limit ECDb versions that can use or import certain things.

Since we are currently unable to increment Read/Write compatible ECDb profile version, we cannot make such changes.
As an alternative we are introducing these custom attributes, so we can introduce new features which require a certain ECDb version to be used/imported.

Basically, trying to query a class flagged as "requires ecdb 4.0.0.6" in an older ecdb version will reject the query, but the file and the remaining contents can still be used normally.

### CODE REVIEW

File: iModelCore/ECDb/ECDb/ProfileUpgrader.cpp
Comment (rschili) [reply iModelCore/ECDb/ECDb/ProfileUpgrader.cpp:18]: Yeah it's strange, but there are others alongside this one in the same file. It's the intended mechanism. This method seems to be for additional code to run during a profile upgrade, but this upgrade only imports an ECDbMap schema, which happens through another mechanism, so there is nothing to do here.

File: iModelCore/ECDb/ECDb/ECSchemas/ECDbMap.02.00.02.ecschema.xml
Comment (pmconne) [inline iModelCore/ECDb/ECDb/ECSchemas/ECDbMap.02.00.02.ecschema.xml:54]: Are these two custom attributes orthogonal? i.e., are there scenarios in which it's ok to import but not to use, or vice versa; or scenarios in which import requires a different profile version than use?

**Review by pmconne — COMMENTED** [review]
Waiting for @khanaffan and @ColinKerr to review.

File: iModelCore/ECDb/ECDb/ECSchemas/ECDbMap.02.00.02.ecschema.xml
Comment (rschili) [reply iModelCore/ECDb/ECDb/ECSchemas/ECDbMap.02.00.02.ecschema.xml:54]: Yes that's the idea. Some concepts can be imported by old runtimes, while others require additional logic during import.
Similarly if a new database imported something and ran the additional logic, it may be fine to use it in an older database.

**Comment by khanaffan** [comment]
@rschili You can add details to CHANGES.MD or put a link there to wiki.

**Comment by rschili** [comment]
> @rschili You can add details to CHANGES.MD or put a link there to wiki.

Done, Updated Changes.md

File: iModelCore/ECDb/ECDb/SchemaManagerDispatcher.cpp
Comment (khanaffan) [inline iModelCore/ECDb/ECDb/SchemaManagerDispatcher.cpp:730]:
```suggestion
if (stmt.Prepare(m_ecdb, SqlPrintfString("SELECT ContainerId FROM main." TABLE_CustomAttribute " WHERE ContainerType = %d AND ClassId = ?", (int)SchemaPersistenceHelper::GeneralizedCustomAttributeContainerType::Class)) != DbResult::BE_SQLITE_OK)
                return;

```

File: iModelCore/ECDb/CHANGES.md
Comment (khanaffan) [inline iModelCore/ECDb/CHANGES.md:11]:
```suggestion
* Two custom attributes were added to the ECDbMap schema, the schema version is incremented to `02.00.02`
```

File: iModelCore/ECDb/Tests/NonPublished/InstanceReaderTest.cpp
Comment (ColinKerr) [inline iModelCore/ECDb/Tests/NonPublished/InstanceReaderTest.cpp:1]: Does this file have any real changes or is this just line ending changes or something similar?

File: iModelCore/ECDb/Tests/NonPublished/ProfileTests.cpp
Comment (ColinKerr) [inline iModelCore/ECDb/Tests/NonPublished/ProfileTests.cpp:853]: how about on struct and relationship class?
How about entity class or relationship class as endpoint of relationship class?
CAs on relationship constraints

**Comment by ColinKerr** [comment]
Overall the changes look ok but some cases need additional testing.

Before this PR can be merged the updated versions of the ECDb schemas must be published in the bis-schemas repo.

This stuff is complex and needs to be properly documented, I think the docs should go into the bis-schemas repo as part of the schema docs.  It may not be the most discoverable place in the itwinjs docs but it will get it on the main site and we can link to it to make it more discoverable.  Docs should have through examples of use.

File: iModelCore/ECDb/Tests/NonPublished/InstanceReaderTest.cpp
Comment (rschili) [reply iModelCore/ECDb/Tests/NonPublished/InstanceReaderTest.cpp:1]: Oh no, this appears to be the same thing Matt had a while back. Yeah the file has real changes, several class ids have been adjusted. You can click the little cogwheel on github and tell it to hide whitespace, then it shows the real diff.
I'll check from which line ending to which this switch was, we probably don't want to push it like this..

Comment (rschili) [reply iModelCore/ECDb/Tests/NonPublished/InstanceReaderTest.cpp:1]: I see what's happening, when I download the file from the main branch through the github website, it has newline line endings. However, when I have my local git switch to the main branch it's using CRLF, for both of the branches here.
Looking at the history, the file was changed from LF to CRLF by @mattbjordan on Aug 3 in Changeset #393 
Then it was changed back from CRLF to LF by Affan on Aug 8 in Changeset #374 (Add support for ECSQL expression tree)

Comment (mattbjordan) [reply iModelCore/ECDb/Tests/NonPublished/InstanceReaderTest.cpp:1]: It's supposed to be CRLF. Most of the files in the repo use CRLF, and any that don't are the outliers. The reverse change keeps getting through in bigger PRs in part because Git doesn't tell you about the line-endings change locally - it only appears on Github. Everyone's VSCode line-endings settings can be different, so maybe look into adding a `.vscode/settings.json` file that helps enforce the correct line-endings? (Probably in a separate PR.) Looks like the setting is:
```json
"files.eol": "\r\n"
```

File: iModelCore/ECDb/Tests/NonPublished/ProfileTests.cpp
Comment (rschili) [reply iModelCore/ECDb/Tests/NonPublished/ProfileTests.cpp:853]: Added tests

**Review by ColinKerr — APPROVED** [review]

**Review by khanaffan — APPROVED** [review]

**Comment by pmconne** [comment]
> Hey @pmconne the two have approved.

I have been waiting for the builds to succeed...the one earlier today failed.

**Review by pmconne — APPROVED** [review]

**Comment by DanRod1999** [comment]
@mergifyio backport release/4.x

**Pull request merged**
